internal/providers/installations.go

```diff
@@ -0,0 +1,101 @@
+//
+// Copyright 2024 Stacklok, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package providers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/rs/zerolog"
+
+	"github.com/stacklok/minder/internal/db"
+	"github.com/stacklok/minder/internal/events"
+)
+
+const (
+	// ProviderInstallationTopic is the topic for when a provider installation is removed
+	ProviderInstallationTopic = "internal.provider.installation.removed.event"
+)
+
+// ProviderInstallationEvent is an event that occurs when a provider installation changes
+type ProviderInstallationEvent string
+
+const (
+	// ProviderInstanceRemovedEvent is an event that occurs when a provider instance is removed
+	ProviderInstanceRemovedEvent ProviderInstallationEvent = "provider_instance_removed"
+)
+
+const (
+	// InstallationEventKey is the key for the event in the message metadata (e.g. removed)
+	InstallationEventKey = "event"
+	// ClassKey is the key for the provider class in the message metadata
+	ClassKey = "class"
+)
+
+// InstallationManager is a struct representing the installation manager
+type InstallationManager struct {
+	svc ProviderService
+}
+
+// NewInstallationManager creates a new installation manager
+func NewInstallationManager(
+	svc ProviderService,
+) *InstallationManager {
+	return &InstallationManager{
+		svc: svc,
+	}
+}
+
+// Register implements the Consumer interface.
+func (im *InstallationManager) Register(reg events.Registrar) {
+	reg.Register(ProviderInstallationTopic, im.handleProviderInstallationEvent)
+}
+
+func (im *InstallationManager) handleProviderInstallationEvent(msg *message.Message) error {
+	ctx := msg.Context()
+	zerolog.Ctx(ctx).Info().Msg("Handling provider installation event")
+
+	event := ProviderInstallationEvent(msg.Metadata.Get(InstallationEventKey))
+	if event == ProviderInstanceRemovedEvent {
+		return im.handleProviderInstanceRemovedEvent(ctx, msg)
+	}
+	zerolog.Ctx(ctx).Error().Msgf("Unknown event: %s", event)
```

Comment:
This feels like it might want to be a `switch` with a `default`, but this is okay for now.